File: DAA/06_quick_sort.cpp
#include <bits/stdc++.h>
using namespace std;
using namespace std::chrono;

// Deterministic Partition for Quick Sort
int partition(vector<int>& arr, int low, int high) {
    int pivot = arr[high];
    int i = low - 1;
    for (int j = low; j <= high - 1; j++) {
        if (arr[j] < pivot) {
            i++;
            swap(arr[i], arr[j]);
        }
    }
    swap(arr[i + 1], arr[high]);
    return i + 1;
}

// Randomized Partition for Quick Sort
int randomPartition(vector<int>& arr, int low, int high) {
    srand(time(NULL));
    int random = low + rand() % (high - low);
    swap(arr[random], arr[high]);
    return partition(arr, low, high);
}

// Deterministic Quick Sort
void deterministicQuickSort(vector<int>& arr, int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);
        deterministicQuickSort(arr, low, pi - 1);
        deterministicQuickSort(arr, pi + 1, high);
    }
}

// Randomized Quick Sort
void randomizedQuickSort(vector<int>& arr, int low, int high) {
    if (low < high) {
        int pi = randomPartition(arr, low, high);
        randomizedQuickSort(arr, low, pi - 1);
        randomizedQuickSort(arr, pi + 1, high);
    }
}

int main() {
    const int SIZE = 100000;
    vector<int> arr(SIZE);

    // Generate random integers
    for (int i = 0; i < SIZE; i++) {
        arr[i] = rand() % 10000;
    }

    // Clone array for both variants
    vector<int> deterministicArr = arr;
    vector<int> randomizedArr = arr;

    // Measure time for deterministic quicksort
    auto startDeterministic = high_resolution_clock::now();
    deterministicQuickSort(deterministicArr, 0, SIZE - 1);
    auto stopDeterministic = high_resolution_clock::now();
    auto deterministicDuration = duration_cast<milliseconds>(stopDeterministic - startDeterministic);

    // Measure time for randomized quicksort
    auto startRandomized = high_resolution_clock::now();
    randomizedQuickSort(randomizedArr, 0, SIZE - 1);
    auto stopRandomized = high_resolution_clock::now();
    auto randomizedDuration = duration_cast<milliseconds>(stopRandomized - startRandomized);

    cout << "Deterministic Quick Sort took " << deterministicDuration.count() << " milliseconds." << endl;
    cout << "Randomized Quick Sort took " << randomizedDuration.count() << " milliseconds." << endl;

    return 0;
}

File: DAA/06_quick_sort.py
import random
import time

# Deterministic Partition for Quick Sort
def partition(arr, low, high):
    pivot = arr[high]
    i = low - 1
    for j in range(low, high):
        if arr[j] < pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1

# Randomized Partition for Quick Sort
def random_partition(arr, low, high):
    random_index = random.randint(low, high)
    arr[random_index], arr[high] = arr[high], arr[random_index]
    return partition(arr, low, high)

# Deterministic Quick Sort
def deterministic_quick_sort(arr, low, high):
    if low < high:
        pi = partition(arr, low, high)
        deterministic_quick_sort(arr, low, pi - 1)
        deterministic_quick_sort(arr, pi + 1, high)

# Randomized Quick Sort
def randomized_quick_sort(arr, low, high):
    if low < high:
        pi = random_partition(arr, low, high)
        randomized_quick_sort(arr, low, pi - 1)
        randomized_quick_sort(arr, pi + 1, high)

if __name__ == "__main__":
    SIZE = 100000
    arr = [random.randint(1, 10000) for _ in range(SIZE)]

    # Clone array for both variants
    deterministic_arr = arr[:]
    randomized_arr = arr[:]

    # Measure time for deterministic quicksort
    start_deterministic = time.time()
    deterministic_quick_sort(deterministic_arr, 0, SIZE - 1)
    end_deterministic = time.time()
    deterministic_duration = (end_deterministic - start_deterministic) * 1000  # Convert to milliseconds

    # Measure time for randomized quicksort
    start_randomized = time.time()
    randomized_quick_sort(randomized_arr, 0, SIZE - 1)
    end_randomized = time.time()
    randomized_duration = (end_randomized - start_randomized) * 1000  # Convert to milliseconds

    print(f"Deterministic Quick Sort took {deterministic_duration} milliseconds.")
    print(f"Randomized Quick Sort took {randomized_duration} milliseconds.")
